### DISCUSSION TITLE
Build System Clarification

### DISCUSSION THREAD

**Discussion by satzguyz** [nasa/fprime#1370] [Q&A]
Hello,

Where can I get the details about the Build system used in the FPrime software. As I see, it has cmake, FPP, SBT, CI, CFDP, python based utilities, etc.

Just confused with the complex structure of the codebase to generate required items.

**Comment by LeStarch** [comment]
The build system for F´  is based on CMake, which calls several tools as part of its process: FPP, and python utilities. These tools are called to generate code from the model files as F´ uses model-driven engineering supported by auto-coding.

The other piece of the build system is `fprime-util`.  This is a helper used to make working with large F´deployments easier.  Typical F´ projects have a development process based around engineering individual components at a time, while the build of the software is deployment focused.  `fprime-util` enables users to build/test in a component directory by mapping their commands to a deployment directory.  It is optional for those who prefer straight CMake.

**Example: fprime-util developer's pattern**
```
cd Ref
fprime-util generate  # Generates a deployment build for Ref
cd SignalGen
fprime-util build   # Builds just "SignalGen" using the reference deployment 
```

**Example: CMake native**
```
cd Ref
mkdir build
cd build
cmake ..  # Generate deployment build for Ref
make SignalGen  # Builds just "SignalGen" using the reference deployment (Note: we are inside the build cache, not in the SignalGen folder)
```

SBT builds FPP.  It will be removed from the F´ build system in the next release.
CI is used to test F´ automatically, and is not required for users.  It runs as part of this repo.
CFDP is an F´ component that may be included in your software.
